src/pygeoprocessing/routing/watershed.pyx

```diff
@@ -470,100 +473,105 @@ cdef cset[CoordinatePair] _c_split_geometry_into_seeds(
-        block_xoff = block_info['xoff']
-        block_yoff = block_info['yoff']
-        n_rows = seed_array.shape[0]
-        n_cols = seed_array.shape[1]
-
-        for row in range(n_rows):
-            for col in range(n_cols):
-                with cython.boundscheck(False):
-                    # Check if the pixel does not overlap the geometry.
-                    if seed_array[row, col] == 0:
-                        continue
-
-                global_row = seed_raster_origin_row + block_yoff + row
-                global_col = seed_raster_origin_col + block_xoff + col
-
-                if write_diagnostic_vector == 1:
-                    diagnostic_layer.StartTransaction()
-                    new_feature = ogr.Feature(diagnostic_layer.GetLayerDefn())
-                    new_feature.SetGeometry(ogr.CreateGeometryFromWkb(
-                        shapely.geometry.Point(
-                            x_origin + ((global_col*x_pixelwidth) +
-                                        (x_pixelwidth / 2.)),
-                            y_origin + ((global_row*y_pixelwidth) +
-                                        (y_pixelwidth / 2.))).wkb))
-                    diagnostic_layer.CreateFeature(new_feature)
-                    diagnostic_layer.CommitTransaction()
-
-                seed_set.insert(CoordinatePair(global_col, global_row))
-
-    return seed_set
+
+    # It's possible for a perfectly vertical or horizontal line to cover 0 rows
+    # or columns, so defaulting to row/col count of 1 in these cases.
+    local_n_cols = max(abs(maxx_aligned - minx_aligned) // abs(x_pixelwidth), 1)
+    local_n_rows = max(abs(maxy_aligned - miny_aligned) // abs(y_pixelwidth), 1)
+
+    with GDALUseExceptions():
+        # The geometry does not fit into a single pixel, so let's create a new
+        # raster onto which to rasterize it.
+        memory_driver = gdal.GetDriverByName('Memory')
+        new_vector = memory_driver.Create('mem', 0, 0, 0, gdal.GDT_Unknown)
+        new_layer = new_vector.CreateLayer('user_geometry', flow_dir_srs, ogr.wkbUnknown)
+        new_layer.StartTransaction()
+        new_feature = ogr.Feature(new_layer.GetLayerDefn())
+        new_feature.SetGeometry(ogr.CreateGeometryFromWkb(source_geom_wkb))
+        new_layer.CreateFeature(new_feature)
+        new_layer.CommitTransaction()
+
+        local_origin_x = max(minx_aligned, x_origin)
+        local_origin_y = min(maxy_aligned, y_origin)
+
+        local_geotransform = [
+            local_origin_x, flow_dir_geotransform[1], flow_dir_geotransform[2],
+            local_origin_y, flow_dir_geotransform[4], flow_dir_geotransform[5]]
+        gtiff_driver = gdal.GetDriverByName('GTiff')
+        # Raster is sparse, no need to fill.
+        raster = gtiff_driver.Create(
+            target_raster_path, int(local_n_cols), int(local_n_rows), 1,
+            gdal.GDT_Byte, options=GTIFF_CREATION_OPTIONS)
+
+        raster.SetSpatialRef(flow_dir_srs)

```

Comment:
It seems like `SetSpatialRef` is equivalent to `SetProjection` but takes in a spatial reference object rather than a WKT string. Calling `ExportToWKT` on an empty spatial reference raises an unhelpful error, so I changed this to avoid the need for it.